README.md
# flutter_steps_animation

[中文文档](README_CN.md)

A powerful Flutter package for building complex animation step by step.

## Example

![example](example.gif)

## Depend on it

Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  flutter_steps_animation: ^1.1.0
```

## Getting Started

Using StepsAnimation is more of a normal Animation:

```dart
class AnimationPageState extends State<AnimationPage>
    with SingleTickerProviderStateMixin {
  StepsAnimation stepsAnimation;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        color: Colors.white,
        child: GestureDetector(
          onTap: () {
            _playAnimation(stepsAnimation.controller);
          },
          child: Center(
            child: AnimatedBuilder(
              animation: stepsAnimation.controller,
              builder: stepsAnimation.builder,
            ),
          ),
        ),
      ),
    );
  }

  @override
  void initState() {
    super.initState();
    stepsAnimation = StepsAnimationBuilder()
        .addStepBuilder(_timeAnimation(1))
        .addStepBuilder(_timeAnimation(2))
        .addStepBuilder(_timeAnimation(3))
        .addStepBuilder(_multipleAnimation())
        .addStepBuilder(_waitStep(2))
        .addStepBuilder(_moreStepsAnimation())
        .animation(this);
  }

  Future<Null> _playAnimation(AnimationController controller) async {
    try {
      await controller.forward().orCancel;
      await controller.reverse().orCancel;
//      controller.reset();
    } on TickerCanceled {}
  }

  @override
  void dispose() {
    stepsAnimation.controller.dispose();
    super.dispose();
  }

```

Add steps for StepsAnimation

```dart
stepsAnimation = StepsAnimationBuilder()
        .addStepBuilder(_timeAnimation(1))
        .addStepBuilder(_timeAnimation(2))
        .addStepBuilder(_timeAnimation(3))
        .addStepBuilder(_multipleAnimation())
        .addStepBuilder(_waitStep(2))
        .addStepBuilder(_moreStepsAnimation())
        .animation(this);
```

Whole code of animations

```dart
SingleAnimationBuilder _timeAnimation(int number) {
    final duration = Duration(milliseconds: 500);
    return SingleAnimationBuilder(
      duration: duration,
      buildInfo: SingleAnimationBuildInfo(
          animatable: Tween<double>(begin: 25, end: 100),
          from: Duration.zero,
          duration: duration),
      buildAnimation: (context, data) {
        return Text(
          '$number',
          style: TextStyle(fontSize: data.value),
        );
      },
    );
  }

  MultipleAnimationBuilder _multipleAnimation() {
    final duration = Duration(seconds: 10);
    final builder = MultipleAnimationBuilder(
        duration: duration,
        buildAnimation: (context, map) {
          final Color color = map['color'].value;
          final complementaryColor = Color(0xffffff ^ color.value);
          return ClipRRect(
            borderRadius:
                BorderRadius.all(Radius.circular(map['radius'].value)),
            child: Container(
              height: map['height'].value,
              width: map['width'].value,
              color: color,
              child: Center(
                child: Text(
                  '${map['color'].value}',
                  style: TextStyle(color: complementaryColor),
                ),
              ),
            ),
          );
        });
    builder
        .addAnimatable(
            animatable: Tween<double>(begin: 100, end: 400),
            from: Duration.zero,
            duration: Duration(seconds: 4),
            key: 'height')
        .addAnimatable(
            animatable: Tween<double>(begin: 400, end: 100),
            from: Duration(seconds: 5),
            duration: Duration(seconds: 5),
            key: 'height')
        .addAnimatable(
            animatable: Tween<double>(begin: 400, end: 150),
            from: Duration.zero,
            duration: Duration(seconds: 4),
            key: 'width')
        .addAnimatable(
            animatable: Tween<double>(begin: 150, end: 400),
            from: Duration(seconds: 5),
            duration: Duration(seconds: 3),
            key: 'width')
        .addAnimatable(
            animatable: ColorTween(
                begin: Colors.green[400], end: Colors.yellowAccent[700]),
            from: Duration(seconds: 2),
            duration: Duration(seconds: 4),
            key: 'color')
        .addAnimatable(
            animatable: ColorTween(
                begin: Colors.yellowAccent[700], end: Colors.red[800]),
            from: Duration(seconds: 7),
            duration: Duration(seconds: 3),
            key: 'color')
        .addAnimatable(
            animatable: Tween<double>(begin: 0, end: 200),
            from: Duration(seconds: 3),
            duration: Duration(seconds: 2),
            key: 'radius')
        .addAnimatable(
            animatable: Tween<double>(begin: 200, end: 0),
            from: Duration(seconds: 6),
            duration: Duration(seconds: 3),
            key: 'radius');
    return builder;
  }

  NoneAnimationBuilder _waitStep(int second) {
    return NoneAnimationBuilder(
        duration: Duration(seconds: second),
        builder: (context) {
          return Container(
            color: Colors.black,
            height: 100,
            width: 400,
            child: Center(
              child: Text(
                'just wait ${second}s',
                style: TextStyle(color: Colors.white),
              ),
            ),
          );
        });
  }

  SingleAnimationBuilder _moreStepsAnimation() {
    return SingleAnimationBuilder(
      duration: Duration(seconds: 10),
      buildAnimation: (context, animation) {
        return Container(
          color: Colors.black,
          height: animation.value,
          width: 400,
          child: Center(
            child: Text(
              'more steps',
              style: TextStyle(color: Colors.white),
            ),
          ),
        );
      },
      buildInfo: SingleAnimationBuildInfo(
        animatable: Tween<double>(begin: 100, end: 1000),
        from: Duration.zero,
        duration: Duration(seconds: 8),
      ),
    );
  }
```





CHANGELOG.md
## [1.1.0] - 2019/12/17

* Modify the implementation of multiple_animation.

## [1.0.2] - 2019/11/29

* Modify Readme.

## [1.0.1] - 2019/11/29

* Modify Readme and example.

## [1.0.0] - 2019/11/28

* First version.


lib/src/base.dart
import 'package:flutter/widgets.dart';

abstract class StepsAnimation {
  final AnimationController controller;

  StepsAnimation(this.controller);

  TransitionBuilder get builder;
}

abstract class OneStepAnimationData<T> {
  final Duration duration;

  final T animationData;

  OneStepAnimationData(this.duration, this.animationData);
}

typedef OneStepBuildAnimation<T> = Widget Function(
    BuildContext context, T data);

abstract class OneStepAnimation<T, AnimationData extends OneStepAnimationData<T>> {
  final AnimationData data;
  final OneStepBuildAnimation<T> oneStepBuildAnimation;

  OneStepAnimation(this.data, this.oneStepBuildAnimation);

  TransitionBuilder get builder => (BuildContext context, Widget child) {
        return oneStepBuildAnimation(context, data.animationData);
      };
}

abstract class OneStepAnimationBuilder {
  final Duration duration;

  OneStepAnimationBuilder(this.duration);

  OneStepAnimation buildStep(AnimationController controller,
      Duration startDuration, Duration totalDuration);
}


lib/src/none_animation.dart
import 'package:flutter/animation.dart';
import 'package:flutter/widgets.dart';

import 'base.dart';

class NoneAnimation extends OneStepAnimation<dynamic, NoneAnimationData> {
  NoneAnimation(NoneAnimationData animationData,
      OneStepBuildAnimation<NoneAnimationData> oneStepBuildAnimation)
      : super(animationData, oneStepBuildAnimation);
}

class NoneAnimationData extends OneStepAnimationData {
  NoneAnimationData(Duration duration) : super(duration, null);
}

class NoneAnimationBuilder extends OneStepAnimationBuilder {
  WidgetBuilder builder;

  NoneAnimationBuilder({@required Duration duration, @required this.builder})
      : super(duration);

  @override
  OneStepAnimation buildStep(AnimationController controller,
      Duration startDuration, Duration totalDuration) {
    final animationData = NoneAnimationData(duration);
    return NoneAnimation(animationData, (context, dynamic _) {
      return builder(context);
    });
  }
}


lib/src/steps_animation.dart
import 'package:flutter/widgets.dart';

import 'base.dart';

class _StepsAnimationImpl extends StepsAnimation {
  final List<OneStepAnimation> stepAnimations;
  final List<double> stepDurationCostingList;

  _StepsAnimationImpl._internal(AnimationController controller,
      this.stepAnimations, this.stepDurationCostingList)
      : super(controller);

  @override
  get builder => (BuildContext context, Widget child) {
        return getCurrentStep().builder(context, child);
      };

  //todo This method assumes that controller's lowerBound and upperBound is 0 and 1
  OneStepAnimation getCurrentStep() {
    final value = controller.value;
    for (int i = 0; i < stepDurationCostingList.length; i++) {
      if (stepDurationCostingList[i] > value) {
        return stepAnimations[i - 1];
      }
    }
    return stepAnimations.last;
  }
}

class StepsAnimationBuilder {
  final List<OneStepAnimationBuilder> stepBuilders = [];

  StepsAnimationBuilder addStepBuilder(OneStepAnimationBuilder stepBuilder) {
    stepBuilders.add(stepBuilder);
    return this;
  }

  _StepsAnimationImpl animation(TickerProvider vsync) {
    final duration = _calculateTotalDuration();
    //todo lowerBound and upperBound must be 0 and 1.
    final controller = AnimationController(
        vsync: vsync, duration: duration, lowerBound: 0, upperBound: 1);
    final stepAnimations = <OneStepAnimation>[];
    final stepDurationCostingList = <double>[];
    Duration stepStartDuration = Duration.zero;
    stepDurationCostingList.add(0);
    stepBuilders.forEach((stepBuilder) {
      final step =
          stepBuilder.buildStep(controller, stepStartDuration, duration);
      stepStartDuration += stepBuilder.duration;
      stepDurationCostingList.add(stepStartDuration.inMilliseconds.toDouble() /
          duration.inMilliseconds.toDouble());
      stepAnimations.add(step);
    });
    return _StepsAnimationImpl._internal(
        controller, stepAnimations, stepDurationCostingList);
  }

  Duration _calculateTotalDuration() {
    Duration duration = Duration.zero;
    stepBuilders.forEach((step) {
      duration += step.duration;
    });
    return duration;
  }
}


lib/src/single_animation.dart
import 'package:flutter/widgets.dart';

import 'base.dart';

class SingleAnimation<T extends Animation>
    extends OneStepAnimation<T, SingleAnimationData<T>> {
  SingleAnimation(SingleAnimationData<T> data, oneStepBuildAnimation)
      : super(data, oneStepBuildAnimation);
}

class SingleAnimationData<T extends Animation> extends OneStepAnimationData<T> {
  SingleAnimationData(
      Duration duration,
      Duration startDuration,
      Duration totalDuration,
      AnimationController controller,
      SingleAnimationBuildInfo info)
      : super(duration,
            _createAnimation(startDuration, totalDuration, controller, info));

  static Animation _createAnimation(
      Duration startDuration,
      Duration totalDuration,
      AnimationController controller,
      SingleAnimationBuildInfo info) {
    final startInMilliseconds = startDuration.inMilliseconds;
    final totalInMilliseconds = totalDuration.inMilliseconds;
    final double begin =
        (startInMilliseconds + info.from.inMilliseconds).toDouble() /
            totalInMilliseconds;
    final double end =
        begin + info.duration.inMilliseconds.toDouble() / totalInMilliseconds;
    CurveTween interval = CurveTween(curve: Interval(begin, end));
    final animatable = info.animatable.chain(interval);
    final animation = animatable.animate(controller);
    return animation;
  }
}

class SingleAnimationBuildInfo {
  final Animatable animatable;
  final Duration from;
  final Duration duration;

  SingleAnimationBuildInfo({
    @required this.animatable,
    @required this.from,
    @required this.duration,
  });
}

class SingleAnimationBuilder<T extends Animation>
    extends OneStepAnimationBuilder {
  final SingleAnimationBuildInfo buildInfo;
  final OneStepBuildAnimation<T> buildAnimation;

  SingleAnimationBuilder({
    @required Duration duration,
    @required this.buildAnimation,
    @required this.buildInfo,
  }) : super(duration);

  @override
  OneStepAnimation<T, SingleAnimationData<T>> buildStep(
      AnimationController controller,
      Duration startDuration,
      Duration totalDuration) {
    final data = SingleAnimationData<T>(
        duration, startDuration, totalDuration, controller, buildInfo);
    return SingleAnimation<T>(data, buildAnimation);
  }
}


lib/src/multiple_animation.dart
import 'package:flutter/animation.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

import 'base.dart';

class _MultipleAnimationBuildInfo {
  final Animatable animatable;
  final Duration from;
  final Duration duration;
  final Object key;

  _MultipleAnimationBuildInfo(
      this.animatable, this.from, this.duration, this.key);
}

class MultipleAnimationBuilder extends OneStepAnimationBuilder {
  final List<_MultipleAnimationBuildInfo> buildInfoList = [];
  final OneStepBuildAnimation<Map<Object, Animation>> buildAnimation;

  MultipleAnimationBuilder({
    @required Duration duration,
    @required this.buildAnimation,
  })  : assert(duration != null && duration != Duration.zero),
        assert(buildAnimation != null),
        super(duration);

  MultipleAnimationBuilder addAnimatable({
    @required Animatable animatable,
    @required Duration from,
    @required Duration duration,
    @required key,
  }) {
    assert(animatable != null);
    assert(from >= Duration.zero);
    assert(from + duration <= this.duration);
    assert(duration > Duration.zero);
    buildInfoList
        .add(_MultipleAnimationBuildInfo(animatable, from, duration, key));
    return this;
  }

  @override
  MultipleAnimation buildStep(AnimationController controller,
      Duration startDuration, Duration totalDuration) {
    final data = MultipleAnimationData(
        duration, startDuration, totalDuration, controller, buildInfoList);
    final multipleStepAnimation = MultipleAnimation(data, buildAnimation);
    return multipleStepAnimation;
  }
}

class MultipleAnimation
    extends OneStepAnimation<Map<Object, Animation>, MultipleAnimationData> {
  MultipleAnimation(MultipleAnimationData data,
      OneStepBuildAnimation<Map<Object, Animation>> oneStepBuildAnimation)
      : super(data, oneStepBuildAnimation);
}

class MultipleAnimationData
    extends OneStepAnimationData<Map<Object, Animation>> {
  MultipleAnimationData(
      Duration duration,
      Duration startDuration,
      Duration totalDuration,
      AnimationController controller,
      List<_MultipleAnimationBuildInfo> list)
      : super(
            duration,
            _createAnimationData(
                startDuration, totalDuration, controller, list));

  static Map<Object, Animation> _createAnimationData(
      Duration startDuration,
      Duration totalDuration,
      AnimationController controller,
      List<_MultipleAnimationBuildInfo> list) {
    final Map<Object, _StartTimeSortedAnimatable> animatableMap = {};
    final startInMilliseconds = startDuration.inMilliseconds;
    final totalInMilliseconds = totalDuration.inMilliseconds;
    list.forEach((_MultipleAnimationBuildInfo multipleInfo) {
      final double begin =
          (startInMilliseconds + multipleInfo.from.inMilliseconds).toDouble() /
              totalInMilliseconds;
      final double end = begin +
          multipleInfo.duration.inMilliseconds.toDouble() / totalInMilliseconds;
      CurveTween interval = CurveTween(curve: Interval(begin, end));

      final animatable = multipleInfo.animatable.chain(interval);
      final key = multipleInfo.key;
      if (animatableMap[key] == null) {
        animatableMap[key] = _StartTimeSortedAnimatable()
          ..addAnimatable(animatable, begin, end, key);
      } else {
        animatableMap[key].addAnimatable(animatable, begin, end, key);
      }
    });
    final Map<Object, Animation> map = {};
    animatableMap.forEach((key, animatable) {
      map[key] = animatable.animate(controller);
    });
    return map;
  }
}

class _AnimatableAndStartTime {
  final Animatable animatable;
  final double begin;
  final double end;

  _AnimatableAndStartTime(this.animatable, this.begin, this.end);
}

class _StartTimeSortedAnimatable extends Animatable {
  final animatableList = <_AnimatableAndStartTime>[];

  void addAnimatable(Animatable animatable, double begin, double end, Object key) {
    animatableList.add(_AnimatableAndStartTime(animatable, begin, end));
    animatableList.sort((a, b) {
      return a.begin.compareTo(b.begin);
    });
    assert(() {
      _AnimatableAndStartTime preAnimatable;
      for (final animatable in animatableList) {
        if (preAnimatable == null) {
          preAnimatable = animatable;
          continue;
        }
        if (preAnimatable.end > animatable.begin) {
          throw 'Animatable of \'$key\' conflicts';
        }
        preAnimatable = animatable;
      }
      return true;
    }());
  }

  Animatable getAppropriateAnimatable(double t) {
    assert(animatableList.length != 0);
    for (int i = 0; i < animatableList.length; ++i) {
      switch (t.compareTo(animatableList[i].begin)) {
        case -1:
          int index = i - 1 >= 0 ? i - 1 : 0;
          return animatableList[index].animatable;
          break;
        case 0:
          return animatableList[i].animatable;
          break;
        case 1:
          continue;
          break;
      }
    }
    return animatableList.last.animatable;
  }

  @override
  transform(double t) {
    final currentAnimatable = getAppropriateAnimatable(t);
    return currentAnimatable.transform(t);
  }
}


example/lib/animation_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_steps_animation/flutter_steps_animation.dart';

class AnimationPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => AnimationPageState();
}

class AnimationPageState extends State<AnimationPage>
    with SingleTickerProviderStateMixin {
  StepsAnimation stepsAnimation;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        color: Colors.white,
        child: GestureDetector(
          onTap: () {
            _playAnimation(stepsAnimation.controller);
          },
          child: Center(
            child: AnimatedBuilder(
              animation: stepsAnimation.controller,
              builder: stepsAnimation.builder,
            ),
          ),
        ),
      ),
    );
  }

  @override
  void initState() {
    super.initState();
    stepsAnimation = StepsAnimationBuilder()
        .addStepBuilder(_timeAnimation(1))
        .addStepBuilder(_timeAnimation(2))
        .addStepBuilder(_timeAnimation(3))
        .addStepBuilder(_multipleAnimation())
        .addStepBuilder(_waitStep(2))
        .addStepBuilder(_moreStepsAnimation())
        .animation(this);
  }

  Future<Null> _playAnimation(AnimationController controller) async {
    try {
      await controller.forward().orCancel;
      await controller.reverse().orCancel;
//      controller.reset();
    } on TickerCanceled {}
  }

  @override
  void dispose() {
    stepsAnimation.controller.dispose();
    super.dispose();
  }

  SingleAnimationBuilder _timeAnimation(int number) {
    final duration = Duration(milliseconds: 500);
    return SingleAnimationBuilder(
      duration: duration,
      buildInfo: SingleAnimationBuildInfo(
          animatable: Tween<double>(begin: 25, end: 100),
          from: Duration.zero,
          duration: duration),
      buildAnimation: (context, data) {
        return Text(
          '$number',
          style: TextStyle(fontSize: data.value),
        );
      },
    );
  }

  //Notice: When multiple animations applied to the same key, the latters will override those former ones.
  //It's your responsibility to maintain the continuity between these animations,
  //or the result effect might be jumping around when you have overlapping or breaking timeline.
  MultipleAnimationBuilder _multipleAnimation() {
    final duration = Duration(seconds: 10);
    final builder = MultipleAnimationBuilder(
        duration: duration,
        buildAnimation: (context, map) {
          final Color color = map['color'].value;
          final complementaryColor = Color(0xffffff ^ color.value);
          return ClipRRect(
            borderRadius:
                BorderRadius.all(Radius.circular(map['radius'].value)),
            child: Container(
              height: map['height'].value,
              width: map['width'].value,
              color: color,
              child: Center(
                child: Text(
                  '${map['color'].value}',
                  style: TextStyle(color: complementaryColor),
                ),
              ),
            ),
          );
        });
    builder
        .addAnimatable(
            animatable: Tween<double>(begin: 100, end: 400),
            from: Duration.zero,
            duration: Duration(seconds: 4),
            key: 'height')
        .addAnimatable(
            animatable: Tween<double>(begin: 400, end: 100),
            from: Duration(seconds: 5),
            duration: Duration(seconds: 5),
            key: 'height')
        .addAnimatable(
            animatable: Tween<double>(begin: 400, end: 150),
            from: Duration.zero,
            duration: Duration(seconds: 4),
            key: 'width')
        .addAnimatable(
            animatable: Tween<double>(begin: 150, end: 400),
            from: Duration(seconds: 5),
            duration: Duration(seconds: 3),
            key: 'width')
        .addAnimatable(
            animatable: ColorTween(
                begin: Colors.green[400], end: Colors.yellowAccent[700]),
            from: Duration(seconds: 2),
            duration: Duration(seconds: 4),
            key: 'color')
        .addAnimatable(
            animatable: ColorTween(
                begin: Colors.yellowAccent[700], end: Colors.red[800]),
            from: Duration(seconds: 7),
            duration: Duration(seconds: 3),
            key: 'color')
        .addAnimatable(
            animatable: Tween<double>(begin: 0, end: 200),
            from: Duration(seconds: 3),
            duration: Duration(seconds: 2),
            key: 'radius')
        .addAnimatable(
            animatable: Tween<double>(begin: 200, end: 0),
            from: Duration(seconds: 6),
            duration: Duration(seconds: 3),
            key: 'radius');
    return builder;
  }

  NoneAnimationBuilder _waitStep(int second) {
    return NoneAnimationBuilder(
        duration: Duration(seconds: second),
        builder: (context) {
          return Container(
            color: Colors.black,
            height: 100,
            width: 400,
            child: Center(
              child: Text(
                'just wait ${second}s',
                style: TextStyle(color: Colors.white),
              ),
            ),
          );
        });
  }

  SingleAnimationBuilder _moreStepsAnimation() {
    return SingleAnimationBuilder(
      duration: Duration(seconds: 10),
      buildAnimation: (context, animation) {
        return Container(
          color: Colors.black,
          height: animation.value,
          width: 400,
          child: Center(
            child: Text(
              'more steps',
              style: TextStyle(color: Colors.white),
            ),
          ),
        );
      },
      buildInfo: SingleAnimationBuildInfo(
        animatable: Tween<double>(begin: 100, end: 1000),
        from: Duration.zero,
        duration: Duration(seconds: 8),
      ),
    );
  }
}
